src/qforte/utils/qubit_tapering.py

```diff
@@ -0,0 +1,345 @@
+
+    ## Find column echelon form (CEF) of augmented matix.
+
+    ### We only work with the columns of the parity check matrix that have not been transformed yet to CEF.
+    ### This is controlled by the "idx" variable
+    idx = 0
+
+    for row in range(n_strings):
+        ### Find a non-zero row
+        nonzero_indices = np.argwhere(prt_chck_mtrx_aug[row,idx:])
+        if nonzero_indices.shape[0] > 0:
+            column_to_be_CEFed = int(nonzero_indices[0]) + idx
+            ### Bring first non-zero element in this row immediately to the right of that of the previous row
+            prt_chck_mtrx_aug[:, [column_to_be_CEFed , idx]] = prt_chck_mtrx_aug[:, [idx, column_to_be_CEFed]]
+            ### Eliminate remaining non-zero row elements in the part of the parity check matrix that is not in CEF
+            for i in range(1,nonzero_indices.shape[0]):
+                column_not_in_CEF = int(nonzero_indices[i]) + idx
+                prt_chck_mtrx_aug[:,column_not_in_CEF] = np.bitwise_xor(prt_chck_mtrx_aug[:, idx], prt_chck_mtrx_aug[:,column_not_in_CEF])
+            ### decrease the number of columns that need to be checked
+            idx += 1
+
+    ## Find basis vectors of kernel by checking for zero columns in the CEF of the parity check matrix
+
+    dim = False
+    basis = np.zeros((1, 2 * n_qubits), dtype=np.uint8)
+    # The first idx columns are nonzero by construction
+    for column in range(idx, 2 * n_qubits):
+        row = n_strings
+        if not np.any(prt_chck_mtrx_aug[:row, column]):
+            new_basis_vector = prt_chck_mtrx_aug[row:, column].reshape((1,2 * n_qubits))
+            if dim is False:
+                basis = new_basis_vector
+                dim = True
+            else:
+                basis = np.concatenate((basis, new_basis_vector), axis=0)
+
+    # 3) Find maximal Abelian subgroup of the basis of ker{E}
+    # This is done using the symplectic bilinear form B that the binary vector space is equipped with:
+    # B(v,w) = v * B * w.T,
+    # where v and w are binary vectors and B is the matrix of the bilinear form.
+    # B(v,w) = 0 means that v and w commute while B(v,w) = 1 means that they anticommute.
+
+    ## Construct matrix of bilinear form.
+
+    zero_matrix = np.zeros((n_qubits, n_qubits), dtype=np.uint8)
+    identity_matrix = np.identity(n_qubits, dtype=np.uint8)
+    blnr_frm_mtrx = np.block([
+        [zero_matrix    , identity_matrix],
+        [identity_matrix, zero_matrix    ]
+        ])
+
+    ## Construct abelian matrix.
+    ## abelian[i,j] contains the value of the bilinear form betweeen the ith and jth basis vectors of ker{E}
+    ## 0/1 in the ith row and jth column of abelian means that the Pauli strings
+    ## associated with the ith and jth binary vectors commute/anticommute.
+
+    abelian = np.matmul(basis, np.matmul(blnr_frm_mtrx, basis.T))
+    abelian %= 2
+
+    ## Find the maximal abelian subgroup iteratively. This is done as follows:
```

Comment:
I'm not convinced that this is going to get the maximal abelian subgroup. How much looking have you done into algorithms for this? A quick Google search gives me one [alternative](https://math.stackexchange.com/questions/473502/how-to-find-maximal-abelian-subgroups/473921), but I haven't done much thinking about it.

Reply:
I should add that we _could_ save this for a later PR.

Reply:
What does Qiskit do to find the maximal abelian subgroup?